src/lib/storage/index/group_key/group_key_index.cpp

```diff
@@ -14,41 +15,67 @@ size_t GroupKeyIndex::estimate_memory_consumption(ChunkOffset row_count, ChunkOf
 }
 
 GroupKeyIndex::GroupKeyIndex(const std::vector<std::shared_ptr<const BaseSegment>>& segments_to_index)
-    : BaseIndex{get_index_type_of<GroupKeyIndex>()},
-      _indexed_segments(std::dynamic_pointer_cast<const BaseDictionarySegment>(segments_to_index[0])) {
-  Assert(static_cast<bool>(_indexed_segments), "GroupKeyIndex only works with dictionary segments_to_index.");
+    : AbstractIndex{get_index_type_of<GroupKeyIndex>()},
+      _indexed_segment(segments_to_index.empty()  // Empty segment list is illegal
+                           ? nullptr              // but range check needed for accessing the first segment
+                           : std::dynamic_pointer_cast<const BaseDictionarySegment>(segments_to_index[0])) {
+  Assert(static_cast<bool>(_indexed_segment), "GroupKeyIndex only works with dictionary segments_to_index.");
   Assert((segments_to_index.size() == 1), "GroupKeyIndex only works with a single segment.");
 
-  // 1) Initialize the index structures
-  // 1a) Set the index_offset to size of the dictionary + 1 (plus one to mark the ending position)
-  //     and set all offsets to 0
-  _index_offsets = std::vector<size_t>(_indexed_segments->unique_values_count() + 1u, 0u);
-  // 1b) Set the _index_postings to the size of the attribute vector
-  _index_postings = std::vector<ChunkOffset>(_indexed_segments->size());
+  // 1) Initialize the index_offset:
+  //    Set the index_offset to size of the dictionary + 1 (plus one to mark the ending position)
+  //    and set all offsets to 0.
+  //    Using `_index_offsets`, we want to count the occurences of each ValueID of the attribute vector (AV).
+  //    The ValueID for NULL in an AV is the highest available ValueID in the dictionary + 1.
+  //    `unique_values_count` returns the size of dictionary which does not store a ValueID for NULL.
+  //    Therefore we have `unique_values_count` ValueIDs (null value id is not included)
+  //    for which we want to count the occurrences.
+  _index_offsets =
+      std::vector<ChunkOffset>(_indexed_segment->unique_values_count() + 1u /*to mark the ending position */, 0u);
 
   // 2) Count the occurrences of value-ids: Iterate once over the attribute vector (i.e. value ids)
   //    and count the occurrences of each value id at their respective position in the dictionary,
   //    i.e. the position in the _index_offsets
-  resolve_compressed_vector_type(*_indexed_segments->attribute_vector(), [&](auto& attribute_vector) {
+  const auto null_value_id = _indexed_segment->null_value_id();
+  auto null_count = 0u;
+
+  resolve_compressed_vector_type(*_indexed_segment->attribute_vector(), [&](auto& attribute_vector) {
     for (const auto value_id : attribute_vector) {
-      _index_offsets[value_id + 1u]++;
+      if (static_cast<ValueID>(value_id) != null_value_id) {
+        _index_offsets[value_id + 1u]++;
```

Comment:
This is a bit confusing, because until 3), _index_offsets does not actually hold any offsets. How about using `value_histogram` first, moving it into `_value_offset` (renamed from _index_offsets) and then running `std::partial_sum` on it?

Reply:
I aggree, I noticed that too. That would definitely improve legibility and comprehensibility.